boards: arm: Add nucleo_h7a3zi-q,

manoj153 opened the pull request:
BSP for nucleo_h7a3zi-q

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

erwango reviewed (changes requested):
Initial comments on the pure form of this PR.
Please split the initial commit in several:
- First commit: soc (soc/ + dts)
- Second commit: Board*
- Any other change related to driver should go in a dedicated commit

Then please exclude any other commit that aren't necessary to use this board

manoj153 committed:
boards: Add support for nucleo_h7a3zi_q

Introduce this board, rcc config: the SoC uses slightly different
domain naming, eg: `cdppre` is equivalent to `d1ppre`
D1 = CD
D2 = CD<..>2
D3 = SRD

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
drivers: clock_control: stm32h7: Support SoC STM32H7A3XX / STM32H7A3XX-Q

clock requirement

Introduce a new group of clock setting to fit in this series of SoC

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
dts: arm: st: h7a3: add support for stm32h7a3

Introduce device tree support for this family of SoC

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
soc: arm: stm32h7: add stm32h7a3xx support

Basic kconfig config.

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
tests: drivers: STM32: Enabling ADC tests for nucleo_h7a3zi_q

This commit enables test_adc to build and run for this new board

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

erwango reviewed (changes requested):
Thanks for this complete work.
Few comments.
Have you been able to validate all peripherals enabled in board dts ?

manoj153 commented:
> Thanks for this complete work. Few comments. Have you been able to validate all peripherals enabled in board dts ?

@erwango I've tried UART, GPIO, so not **all**

erwango commented:
> @erwango I've tried UART, GPIO, so not all

So please remove the ones you've not tested.

manoj153 commented:
> > @erwango I've tried UART, GPIO, so not all
> 
> So please remove the ones you've not tested.



> > @erwango I've tried UART, GPIO, so not all
> 
> So please remove the ones you've not tested.

Sure.

manoj153 committed:
boards: nucleo h7a3zi-q: enable SMPS

SMPS supply is default for this board

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 commented:
> Related to: #34732
> 
> I am aware that this board's user manual states the default supply is SMPS, so should I add the logic to choose SMPS as the default supply? As far I know, `clock_stm32_ll_h7.c` and `soc_m7.c` are the source to modify. @FRASTM @erwango

I have addressed the SMPS support for this board. https://github.com/zephyrproject-rtos/zephyr/pull/39295/commits/da4360fea75bbb965332f2636231ed78a1e57270
CI is complaining because this PR is dependent on another. https://github.com/zephyrproject-rtos/zephyr/pull/39555

manoj153 commented:
How do we handle cases where SoC has extension Q?
E.g. (STM32H7A3xxQ)

Are they the same entity (STM32H7A3xx) vs. (STM32H7A3xxQ)?
AFAIK, the tree has no such example yet.

 I just realised I need to define STM32H7A3xxQ to include the correct defines(STM32H7A3xxQ), and eventually, it points to the correct header files.

@erwango @FRASTM Have something to say?

erwango commented:
> How do we handle cases where SoC has extension Q? E.g. (STM32H7A3xxQ)
> 
> Are they the same entity (STM32H7A3xx) vs. (STM32H7A3xxQ)? AFAIK, the tree has no such example yet.
> 
> I just realised I need to define STM32H7A3xxQ to include the correct defines(STM32H7A3xxQ), and eventually, it points to the correct header files.
> 
> @erwango @FRASTM Have something to say?

Difference between both CMSIS files is that the Q version will do `#define SMPS`.
Currently in Zephyr, we don't have specific policy around this (one possibility would be to define SMPS support via device tree). Anyway, in any case, it's better to point to correct file.

Btw, what are your plans towards this PR wrt requested changes (ie remove references peripherals that are not tested)?
On my side, I'd prefer that you perform the expected changes (ie get the PR merged with minimal peripheral support). Then you can open a PR with peripheral additions. This way of working would be much easiezr from review perspective so I don't  have to ask the status of the code you're pushing, and reviewing again and again the same code that was initially correct.

manoj153 commented:
> > How do we handle cases where SoC has extension Q? E.g. (STM32H7A3xxQ)
> > Are they the same entity (STM32H7A3xx) vs. (STM32H7A3xxQ)? AFAIK, the tree has no such example yet.
> > I just realised I need to define STM32H7A3xxQ to include the correct defines(STM32H7A3xxQ), and eventually, it points to the correct header files.
> > @erwango @FRASTM Have something to say?
> 
> @manoj153 Difference between both CMSIS files is that the Q version will do `#define SMPS`. Currently in Zephyr, we don't have specific policy around this (one possibility would be to define SMPS support via device tree). Anyway, in any case, it's better to point to correct file.
> 
> Btw, what are your plans towards this PR wrt requested changes (ie remove references peripherals that are not tested)? On my side, I'd prefer that you perform the expected changes (ie get the PR merged with minimal peripheral support). Then you can open a PR with peripheral additions. This way of working would be much easiezr from review perspective so I don't have to ask the status of the code you're pushing, and reviewing again and again the same code that was initially correct.


This PR is hanging due to the SMPS. So the quick way is to introduce the `#define SPMS` from the device tree if we all agree. 


I tested a couple of peripheral that I will share the result. The delay was due to this power supply deadlock issue which I noticed during my test, from my observation with https://github.com/zephyrproject-rtos/zephyr/pull/39555 fixes the problem.

manoj153 commented:
> > > How do we handle cases where SoC has extension Q? E.g. (STM32H7A3xxQ)
> > > Are they the same entity (STM32H7A3xx) vs. (STM32H7A3xxQ)? AFAIK, the tree has no such example yet.
> > > I just realised I need to define STM32H7A3xxQ to include the correct defines(STM32H7A3xxQ), and eventually, it points to the correct header files.
> > > @erwango @FRASTM Have something to say?
> > 
> > 
> > @manoj153 Difference between both CMSIS files is that the Q version will do `#define SMPS`. Currently in Zephyr, we don't have specific policy around this (one possibility would be to define SMPS support via device tree). Anyway, in any case, it's better to point to correct file.
> > Btw, what are your plans towards this PR wrt requested changes (ie remove references peripherals that are not tested)? On my side, I'd prefer that you perform the expected changes (ie get the PR merged with minimal peripheral support). Then you can open a PR with peripheral additions. This way of working would be much easiezr from review perspective so I don't have to ask the status of the code you're pushing, and reviewing again and again the same code that was initially correct.
> 
> @erwango This PR is hanging due to the SMPS. So the quick way is to introduce the `#define SPMS` from the device tree if we all agree. Alternatively, maybe adding logic in the kconfig such that if `CONFIG_POWER_SUPPLY_SMPS` is enabled, set SOC as STM32H7A3xxQ. Let me know which sounds like the best move for this?
> 
> I tested a couple of peripheral that I will share the result. The delay was due to this power supply deadlock issue which I noticed during my test, from my observation with [#39555 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/39555#pullrequestreview-788350760) fixes the problem.

@erwango free to suggest?

erwango commented:
@manoj153 I suggest to get https://github.com/zephyrproject-rtos/zephyr/pull/39555 merged first

FRASTM commented:
LGTM

manoj153 committed:
soc: arm: stm32h7: kconfig workaround to handle SoC H7A3XX-Q P/N

Set the correct soc string so that the correct CMSIS file
is being utilised.

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
samples: drivers: nucleo_h7a3zi_q: Configuring ADC sample application

I tested ADC sample on the board mentioned above
0v = 0
3.3v = 4095

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 commented:
> @manoj153 I suggest to get #39555 merged first

Hi @erwango 

> @manoj153 I suggest to get #39555 merged first

Hi @erwango,  #39555  is merged, so mind taking a look at this PR?

I assume this PR is ready from my side.

erwango commented on boards/arm/nucleo_h7a3zi_q/nucleo_h7a3zi_q.dts:
```diff
@@ -0,0 +1,110 @@
+
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
+	current-speed = <115200>;
```
Would you mind adding:
```
 pinctrl-names = "default";
```
That's not required for now, but soon will be and will avoid conflicts and sync issues with upcoming https://github.com/zephyrproject-rtos/zephyr/pull/40194

erwango commented on boards/arm/nucleo_h7a3zi_q/nucleo_h7a3zi_q.dts:
```diff
@@ -0,0 +1,110 @@
+	pwm12: pwm {
+		status = "okay";
+		st,prescaler = <10000>;
+		pinctrl-0 = <&tim12_ch1_pb14>;
```
Please add:
```
pinctrl-names = "default";
```

erwango commented on boards/arm/nucleo_h7a3zi_q/nucleo_h7a3zi_q.dts:
```diff
@@ -0,0 +1,110 @@
+};
+
+&adc1 {
+	pinctrl-0 = <&adc1_inp15_pa3>;
```
Same here.

erwango commented on boards/arm/nucleo_h7a3zi_q/nucleo_h7a3zi_q.dts:
```diff
@@ -0,0 +1,110 @@
+
+	pwm12: pwm {
+		status = "okay";
+		st,prescaler = <10000>;
```
Please move `st,prescaler` to the parent node. Cf #39628

erwango replied:
This will fix related CI error:
`devicetree error: 'st,prescaler' is marked as deprecated in 'properties:' in /workdir/zephyr/dts/bindings/pwm/st,stm32-pwm.yaml for node /soc/timers@40001800/pwm.`

erwango reviewed (changes requested):
Few more changes and that should be fine.
Thanks for your patience.

tejlmand commented on boards/arm/nucleo_h7a3zi_q/board.cmake:
```diff
@@ -0,0 +1,8 @@
+
+include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
```
Missing newline.

manoj153 replied:
Thanks done

tejlmand commented on boards/arm/nucleo_h7a3zi_q/nucleo_h7a3zi_q_defconfig:
```diff
@@ -25,3 +25,6 @@ CONFIG_GPIO=y
 CONFIG_CLOCK_CONTROL=y
+
+# Enable SMPS
+CONFIG_POWER_SUPPLY_SMPS=y
```
This breaks bisectability in commit: 005b0858b95be96aab6383e2df32b5704658400a
```
$ git checkout 005b0858b95be96aab6383e2df32b5704658400a
$ cmake -GNinja -DBOARD=nucleo_h7a3zi_q -Bbuild samples/hello_world
...
$ ninja -Cbuild
...
[77/149] Building C object zephyr/CMakeFiles/zephyr.dir/soc/arm/st_stm32/stm32h7/soc_m7.c.obj
FAILED: zephyr/CMakeFiles/zephyr.dir/soc/arm/st_stm32/stm32h7/soc_m7.c.obj 
ccache /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc -DBUILD_VERSION=v2.7.99-1511-g005b0858b95b -DCORE_CM7 -DHSE_VALUE=8000000 -DKERNEL -DSTM32H7A3xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I../kernel/include -I../arch/arm/include -I../include -Izephyr/include/generated -I../soc/arm/st_stm32/stm32h7 -I../drivers -I../soc/arm/st_stm32/common -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/soc -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include/Legacy -I/projects/github/ncs/modules/hal/stm32/stm32cube/common_ll/include -isystem ../lib/libc/minimal/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include-fixed -Os -imacros /projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/zephyr.dir/soc/arm/st_stm32/stm32h7/soc_m7.c.obj -MF zephyr/CMakeFiles/zephyr.dir/soc/arm/st_stm32/stm32h7/soc_m7.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/soc/arm/st_stm32/stm32h7/soc_m7.c.obj -c /projects/github/ncs/zephyr/soc/arm/st_stm32/stm32h7/soc_m7.c
/projects/github/ncs/zephyr/soc/arm/st_stm32/stm32h7/soc_m7.c: In function 'stm32h7_init':
/projects/github/ncs/zephyr/soc/arm/st_stm32/stm32h7/soc_m7.c:86:2: error: #error Unsupported configuration: Selected SoC do not support SMPS
   86 | #error Unsupported configuration: Selected SoC do not support SMPS
      |  ^~~~~
[94/149] Building C object zephyr/drivers/timer/CMakeFiles/drivers__timer.dir/cortex_m_systick.c.obj
ninja: build stopped: subcommand failed.
```

Please introduce `CONFIG_POWER_SUPPLY_SMPS=y` together with the SoC supporting this, or in a following commit.

For example by `rebase -i` and re-order the commits.

manoj153 replied:
Ok re-ordered.

tejlmand reviewed (commented):
This PR contains commits that cannot compile, for example 005b0858b95be96aab6383e2df32b5704658400a but also this one 92ff7672e1b302005fd65fd3bcdac45389eedf3b:
```
$ git checkout 92ff7672e1b302005fd65fd3bcdac45389eedf3b
$ cmake -GNinja -DBOARD=nucleo_h7a3zi_q -Bbuild samples/hello_world/
...
$ ninja -Cbuild
FAILED: zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj 
ccache /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc -DBUILD_VERSION=v2.7.99-1508-g92ff7672e1b3 -DCORE_CM7 -DHSE_VALUE=8000000 -DKERNEL -DSTM32H7A3xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../include -Izephyr/include/generated -I../soc/arm/st_stm32/stm32h7 -I../drivers -I../soc/arm/st_stm32/common -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/soc -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include/Legacy -I/projects/github/ncs/modules/hal/stm32/stm32cube/common_ll/include -isystem ../lib/libc/minimal/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include-fixed -Os -imacros /projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj -MF zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj.d -o zephyr/drivers/clock_control/CMakeFiles/drivers__clock_control.dir/clock_stm32_ll_h7.c.obj -c /projects/github/ncs/zephyr/drivers/clock_control/clock_stm32_ll_h7.c
/projects/github/ncs/zephyr/drivers/clock_control/clock_stm32_ll_h7.c: In function 'stm32_clock_control_init':
/projects/github/ncs/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:507:27: error: 'LL_AHB4_GRP1_PERIPH_HSEM' undeclared (first use in this function); did you mean 'LL_AHB2_GRP1_PERIPH_HSEM'?
  507 |  LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_HSEM);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~
      |                           LL_AHB2_GRP1_PERIPH_HSEM
/projects/github/ncs/zephyr/drivers/clock_control/clock_stm32_ll_h7.c:507:27: note: each undeclared identifier is reported only once for each function it appears in
[136/149] Building C object zephyr/kernel/CMakeFiles/kernel.dir/timer.c.obj
ninja: build stopped: subcommand failed.
```

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.defconfig.stm32h7a3xx:
```diff
@@ -6,7 +6,8 @@
 
 config SOC
-	default "stm32h7a3xx"
+	default "stm32h7a3xxQ" if POWER_SUPPLY_SMPS
+	default "stm32h7a3xx" if !POWER_SUPPLY_SMPS
```
This is conceptually wrong as this is selecting the HW (well CMSIS file description) based on SW configuration.
This link between `POWER_SUPPLY_SMPS` and HW reference should be reworked

manoj153 replied:
The alternate idea will be defining an `SMPS` node on the device tree, and the base SoC is stm32h7a3xx SoC, but then it won't  point to the exact CMIS file

What will be a better way of accomplishing it?

erwango replied:
Ideal way would be device tree configuration, but for now we're missing the proper regulator minimum description to support this (at least for current PR).
Problem comes from `POWER_SUPPLY_SMPS`:
- it is user-selectable
- description: `Enable SMPS Power Supply.`

So either you need to add an additional Kconfig symbol, non user-selectable (promptless), like `HAS_POWER_SUPPLY_SMPS`, which would select the corect CSMSIS file and implies `POWER_SUPPLY_SMPS` either `POWER_SUPPLY_SMPS` definition should be reviewed.

manoj153 replied:
Sorry don't understand the following:

- `HAS_POWER_SUPPLY_SMPS` is set by who and where?

The idea is `POWER_SUPPLY_SMPS` acts as user/board-level decision input to enable/disable the SMPS feature, so what is wrong with the current description?

erwango replied:
The whole system ends up selecting a SoC description (`stm32h7a3xx` or `stm32h7a3xxQ`).
This is wrong, because SoC is not configurable but the base of the target configuration:
The SW configuration should be a consequence of the SoC selection and not the other way round.

manoj153 replied:
> 

Please take a look at my last temp commit. Does that meet all the requirements?

manoj153 replied:
@erwango, mind taking a look?

manoj153 replied:
Hi @erwango, any update regarding this PR?

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.soc:
```diff
@@ -46,4 +47,10 @@ config SOC_STM32H7A3XX
 	select CPU_HAS_FPU_DOUBLE_PRECISION
 
+config SOC_STM32H7A3XXQ
+	bool "STM32H7A3XX"
+	select CPU_CORTEX_M7
+	select CPU_HAS_FPU_DOUBLE_PRECISION
```
```suggestion
	bool "STM32H7A3XXQ"
	select SOC_STM32H7A3XX
```

erwango replied:
Please do this change

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.defconfig.stm32h7a3xx:
```diff
@@ -3,13 +3,26 @@
 config SOC
-	default "stm32h7a3xxQ" if POWER_SUPPLY_SMPS
-	default "stm32h7a3xx" if !POWER_SUPPLY_SMPS
+	default "stm32h7a3xx"
```
In this file, the only change required is this one:
```
	default "stm32h7a3xxQ" if HAS_POWER_SUPPLY_SMPS
	default "stm32h7a3xx" if !HAS_POWER_SUPPLY_SMPS
```

This is correct if `HAS_POWER_SUPPLY_SMPS` isn't user selectable but only indirectly selected by the use of `SOC_STM32H7A3XXQ` in board definition

erwango replied:
Let's keep this part the way it is. This is fine

manoj153 replied:
Do you mean fallback to condition upon `POWER_SUPPLY_SMPS`?

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -113,7 +113,8 @@
 #define AHB_FREQ_MAX		275000000UL
 #define APBx_FREQ_MAX		137500000UL
-#elif defined(CONFIG_SOC_STM32H7A3XX)
+#elif defined(CONFIG_SOC_STM32H7A3XX) ||\
+	  defined(CONFIG_SOC_STM32H7A3XXQ)
```
Only difference between `CONFIG_SOC_STM32H7A3XX` and `CONFIG_SOC_STM32H7A3XX` is SMPS support, so everywhere else, `if defined(CONFIG_SOC_STM32H7A3XX)` should work for both

erwango reviewed (dismissed):
We're getting close. Thanks for this change

erwango commented on soc/arm/st_stm32/Kconfig:
```diff
@@ -14,6 +15,10 @@ config SOC_FAMILY
 	string
 	default "st_stm32"
 
+config HAS_POWER_SUPPLY_SMPS
+	bool
+	default n
+
```
Let's remove this symbol

erwango commented on soc/arm/st_stm32/common/Kconfig.soc:
```diff
@@ -24,6 +25,6 @@ config USE_STM32_ASSERT
 
 config POWER_SUPPLY_SMPS
 	bool "Enable STM32 internal SMPS"
-	depends on SOC_SERIES_STM32H7X
+	depends on SOC_SERIES_STM32H7X && HAS_POWER_SUPPLY_SMPS
```
Please remove this change

erwango reviewed (changes requested):
@manoj153 I'm proposing a last version of the SMPS selection, which aims at being more simple.
Let's remove the use of HAS_POWER_SUPPLY_SMPS and deal with the CMSIS selection directly with the use of SOC_STM32H7A3XXQ as you initially propose in this temp proposal.

Let me know if this is unclear (which I would understand)

manoj153 committed:
soc: arm: stm32h7: add stm32h7a3xxq

The Q variant is the same as non-Q, except the Q has SMPS built-in.
This symbol addition is to have the correct
SOC definition ("STM32H7A3XXQ")

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 commented:
I noticed a recent `POWER_SUPPLY_DIRECT_SMPS` addition. I have updated this PR accordingly.

manoj153 commented:
@erwango facing some Kconfigs issue 

```
warning: the choice symbol SOC_STM32H7A3XX (defined at soc/arm/st_stm32/stm32h7/Kconfig.soc:45) is selected by the following symbols, but select/imply has no effect on choice symbols
 - SOC_STM32H7A3XXQ (defined at soc/arm/st_stm32/stm32h7/Kconfig.soc:50)

warning: CPU_CORTEX_M_HAS_DWT (defined at arch/arm/core/aarch32/cortex_m/Kconfig:89) has direct dependencies !CPU_CORTEX_M0 && !CPU_CORTEX_M0PLUS && !CPU_CORTEX_M1 && CPU_CORTEX_M && ARM with value n, but is currently being y-selected by the following symbols:
 - SOC_SERIES_STM32H7X (defined at soc/arm/st_stm32/stm32h7/Kconfig.series:6), with value y, direct dependencies <choice> (value: y), and select condition <choice> (value: y)

error: Aborting due to Kconfig warnings
```

manoj153 commented:
@erwango hi, will you mind looking at this?

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -534,8 +539,11 @@ static int stm32_clock_control_init(const struct device *dev)
 #if !defined(CONFIG_CPU_CORTEX_M4)
 
 	/* HW semaphore Clock enable */
+#if defined(CONFIG_SOC_STM32H7A3XX)
+	LL_AHB2_GRP1_EnableClock(LL_AHB2_GRP1_PERIPH_HSEM);
+#else
 	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_HSEM);
-
+#endif
```
Please don't delete this empty line.

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.defconfig.stm32h7a3xx:
```diff
@@ -6,7 +6,8 @@
 
 config SOC
-	default "stm32h7a3xx"
+	default "stm32h7a3xxQ" if POWER_SUPPLY_DIRECT_SMPS
+	default "stm32h7a3xx" if !POWER_SUPPLY_DIRECT_SMPS
```
```suggestion
	default "stm32h7a3xxQ" if SOC_STM32H7A3XXQ
	default "stm32h7a3xx" if !SOC_STM32H7A3XXQ
```

manoj153 replied:
@erwango 
I made all the changes as your request, but please note we have a build warning consequently build error,

The reason is the way we select `SOC_STM32H7A3XX` from `SOC_STM32H7A3XXQ`. So what should we do?

```
warning: the choice symbol SOC_STM32H7A3XX (defined at soc/arm/st_stm32/stm32h7/Kconfig.soc:45) is selected by the following symbols, but select/imply has no effect on choice symbols
 - SOC_STM32H7A3XXQ (defined at soc/arm/st_stm32/stm32h7/Kconfig.soc:50)

error: Aborting due to Kconfig warnings
```

erwango replied:
@manoj153 Ok, thanks for having add a try. I also just made some attempts on my side, and I don't find any better way than finally defining 2 different, independent SoC symbols, `SOC_STM32H7A3XX` and `SOC_STM32H7A3XXQ` that will have to live separately.
So let's do this:

```
 config SOC_STM32H723XX
 	bool "STM32H7A3XX"
 	select CPU_CORTEX_M7
 	select CPU_HAS_FPU_DOUBLE_PRECISION
 
 config SOC_STM32H7A3XXQ
 	bool "STM32H7A3XXQ"
	select CPU_CORTEX_M7
	select CPU_HAS_FPU_DOUBLE_PRECISION
```

Then:
```
 if SOC_STM32H7A3XX || SOC_STM32H7A3XXQ
 
 config SOC
	default "stm32h7a3xxQ" if SOC_STM32H7A3XXQ
	default "stm32h7a3xx" if SOC_STM32H7A3XX
 
 config NUM_IRQS
 	default 155
 
 endif # SOC_STM32H7A3XX
```

And finally, use `defined(CONFIG_SOC_STM32H7A3XX) || defined(CONFIG_SOC_STM32H7A3XXQ)`
instead of `defined(CONFIG_SOC_STM32H7A3XX)`.

I hope we can do better once SMPS is moved to device tree.

Anyway, thanks for your patience in doing the multiple tries.

manoj153 replied:
Thanks, @erwango. I have made it as requested.

erwango reviewed (approved):
Thanks @manoj153. Approved now.
Might require some commit reordering to make sure stuff is defined before being used, but I won't block for this, considering the number of previous iterations.

manoj153 commented:
> Thanks @manoj153. Approved now. Might require some commit reordering to make sure stuff is defined before being used, but I won't block for this, considering the number of previous iterations.

Can you let me know which commit are they?

erwango commented:
> Can you let me know which commit are they?

In some commit you use board, while board is not yet introduced. Same for `SOC_STM32H7A3XXQ` symbol.
At least this is the way it is shown in github. Please discard my comment if this is not the same order on your branch (weird but happens some times).

manoj153 commented:
> 

I am looking at it, and I can move the `samples: drivers: nucleo_h7a3zi_q: Configuring ADC sample application` commit after introducing the board (no code-changes) just re-ordering commit

erwango commented:
@manoj153 Would you mind fixing issue reported by CI [here](https://github.com/zephyrproject-rtos/zephyr/runs/4762483128?check_suite_focus=true)

manoj153 commented:
> @manoj153 Would you mind fixing issue reported by CI [here](https://github.com/zephyrproject-rtos/zephyr/runs/4762483128?check_suite_focus=true)

I think it's the due absence of `arduino_i2c`

![image](https://user-images.githubusercontent.com/17961318/148792477-49a6d60d-04c0-4858-8fe7-592da439ae04.png)

erwango commented:
> > @manoj153 Would you mind fixing issue reported by CI [here](https://github.com/zephyrproject-rtos/zephyr/runs/4762483128?check_suite_focus=true)
> 
> I think it's the due absence of `arduino_i2c`
> 
> ![image](https://user-images.githubusercontent.com/17961318/148792477-49a6d60d-04c0-4858-8fe7-592da439ae04.png)
> 
> ```
> devicetree error: /__w/zephyr/zephyr/boards/shields/x_nucleo_iks02a1/x_nucleo_iks02a1.overlay:16 (column 1): parse error: undefined node label 'arduino_i2c'
> ```

Ok, then that's because you mention `arduino_i2c` in board yaml file.
```
supported:
  - arduino_gpio
  - arduino_i2c
```
Otherwise you need to do the full work of defining both `arduino_i2c` AND `i2c1` node, but I advise you NOT to do so now and keep it for a next PR.

manoj153 commented:
Ok, valid. Should I drop arduino_gpio too?

erwango commented:
> Ok, valid. Should I drop arduino_gpio too?

No this one is fine as there is proper arduino gpio table description.

manoj153 commented:
> > Ok, valid. Should I drop arduino_gpio too?
> 
> No this one is fine as there is proper arduino gpio table description.

Thanks, CI is happy now, thanks to you!

manoj153 commented:
@FRASTM @ABOSTM Will, you guys, mind formally doing a review and +1 if it looks good to you. 

Thanks

ABOSTM commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -0,0 +1,71 @@
+		zephyr,memory-region = "SRAM1";
+	};
+
+	/* System data RAM accessible over AXI bus: AXI SRAM1 in CD domain */
```
```suggestion
	/* System data RAM accessible over AXI bus: AXI SRAM3 in CD domain */
```

ABOSTM commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -0,0 +1,71 @@
+		reg = <0x24000000 DT_SIZE_K(256)>;
+	};
+
+	/* System data RAM accessible over AXI bus: AXI SRAM1 in CD domain */
```
```suggestion
	/* System data RAM accessible over AXI bus: AXI SRAM2 in CD domain */
```

ABOSTM commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -0,0 +1,71 @@
+		flash-controller@52002000 {
+			flash0: flash@8000000 {
+				write-block-size = <32>;
+				erase-block-size = <DT_SIZE_K(128)>;
```
Flash sector (erase size) is 8K

manoj153 replied:
@ABOSTM  Thanks, I Fixed all 3.

FRASTM commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -0,0 +1,71 @@
+		};
+
+		dmamux1: dmamux@40020800 {
+			dma-requests= <107>;
```
Actually, the dma-request gives the max DMAREQ_ID[6:0] --> upper limit of DMAMUX1 req inputs table. Here, more than 107 + 8 generator req. entries are available. 
Setting dma-request to 107 plus adding the 8 dmamux generator req gives a lower value to the max possible value of the table. Even some (ex. 113, 114, 115 ) are reserved, other values like 116 to 119 are still  possible. The DMAREQ_ID[6:0] of the DMAMUX1_CxCR is a value from 1 to 119. --> considering 8 dmamux generator entries, the dma-requests is 111.

manoj153 replied:
Do you mean also reserved are excluded?

![image](https://user-images.githubusercontent.com/17961318/149606796-fc7a51c1-f14b-4201-ba65-73d56049fcdb.png)


I can see from the datasheet that the total mux input is `127`, reserved are `16`, which means the max possible ID entry is `111`. I wanted to confirm I am right, and please correct me if anything is wrong. Thanks, and I appreciate your review and feedback.

FRASTM reviewed (approved):
minor rmk else LGTM
